File: code/Python/32-Longest Valid Parentheses.py
# -*- coding: utf-8 -*
"""
@author: Kaka Chen
@license: (C) Copyright Personal Uncommercial Limited.
@contact: kakachen1990@gmail.com
@file: 32-Longest Valid Parentheses.py
@time: 2020/7/4 11:19 上午
@desc:

Given a string containing just the characters '(' and ')', find the length of the longest valid (well-formed) parentheses substring.

Example 1:

Input: "(()"
Output: 2
Explanation: The longest valid parentheses substring is "()"
Example 2:

Input: ")()())"
Output: 4
Explanation: The longest valid parentheses substring is "()()"
"""

class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack = []
        res, last = 0, -1
        n = len(s)
        for i in range(n):
            if s[i] == '(':
                stack.append(i)
            elif s[i] == ')':
                if len(stack) == 0:
                    last = i
                else:
                    stack.pop()
                    if len(stack) == 0:
                        res = max(res, i - last)
                    else:
                        res = max(res, i - stack[-1])
        return res


if __name__ == '__main__':
    inputs = ["(()", "()", "(())"]
    s = Solution()
    for inp in inputs:
        print(s.longestValidParentheses(inp))

File: code/C++/32-Longest Valid Parentheses.cpp
/*Longest Valid Parentheses
Given a string containing just the characters '(' and ')', find the length of the longest valid (well-formed) parentheses substring.

For "(()", the longest valid parentheses substring is "()", which has length = 2.

Another example is ")()())", where the longest valid parentheses substring is "()()", which has length = 4.

维护一个堆栈，我之前大体思路是对的，但是在怎么计算结果上有问题
*/

#include <iostream>
#include <stack>
#include <algorithm>

using namespace std;

 int longestValidParentheses(string s) {
        stack<int> st;
        int max_len=0,last=-1;
        const int len=s.size();

        for(int i=0;i<len;i++){
	       
	       if(s[i]=='(')
	           st.push(i);
	       else if (s[i]==')') {
		      if(st.empty())
		         last=i;//更新找到的最后一个')'位置
		      else{//找到一一对'('
			     st.pop();
			     if(st.empty())
			        max_len=max(max_len,i-last);
			     else 
			        max_len=max(max_len,i-st.top());
		   }
	    }
    } 

    return max_len;
        
    }
int main(int argc, char *argv[]) {
	cout<<longestValidParentheses("(()");
}